message-type: Add support for "direct" as value for `type` parameter.

laurynmm opened the pull request:
For endpoints with a `type` parameter to indicate whether the message is a stream or direct  message, [`POST /typing`](https://zulip.com/api/set-typing-status) and [`POST /messages`](https://zulip.com/api/send-message), adds support for passing `"direct"` as a value for direct messages.

Maintains support for `"private"` as a deprecated alias for direct messages.

Fixes https://github.com/zulip/zulip/issues/24960.

---

**Notes**:
- For the **Usage examples** in the `POST /messages` documentation:
  - Comments now use 'direct messages' or 'DMs'
  - Updated the curl and javascript examples
  - Did not update the python example for "direct". I wasn't sure if we wanted to update that in tandem with an update to zulip-python-api; see question below.
- I added the validation check for the `type` parameter that was in `POST /typing` to also be in `POST /messages`.
- I did the refactor in the prep commit because any follow-up changes we do can be a bit clearer in that all instances of `message_type_name` now refer to these parameters passed to the view functions from clients. See `git-grep` below.
  - for `POST /messages`, we use "private" for the `message_type` string that's then used for the various function calls.
  - for `POST /typing`, we use "direct" for the default and updated `message_type` string.
- Updates these two API endpoint documentation pages to use "direct messages" or "DMs" and also added links to relevant help center documentation.

<details>
<summary>git-grep of `message_type_name` after changes</summary>

```console
$ git grep message_type_name
zerver/tests/test_message_send.py:        message_type_name = ["direct", "private"]
zerver/tests/test_message_send.py:        for message_type in message_type_name:
zerver/tests/test_message_send.py:        message_type_name = ["direct", "private"]
zerver/tests/test_message_send.py:        for message_type in message_type_name:
zerver/tests/test_typing.py:        message_type_name = ["direct", "private"]
zerver/tests/test_typing.py:        for message_type in message_type_name:
zerver/views/message_send.py:    message_type_name: str = REQ("type", str_validator=check_string_in(VALID_MESSAGE_TYPES)),
zerver/views/message_send.py:    message_type = message_type_name
zerver/views/typing.py:    message_type_name: str = REQ(
zerver/views/typing.py:    message_type = message_type_name

```
</details>

**Questions**:
- In the **Changes** notes, I stated that "private" would be removed in a future release, but I wasn't sure if that's the case?
- I added some TODO notes where `message_type` is set in both view functions, but these might not be something we plan on doing?
- Again, as noted above, we'll likely want to update the [zulip-python-api](https://github.com/zulip/python-zulip-api/blob/02586f1d348dbcab3bc8d30b805fffd48a524e18/zulip/zulip/send.py#L89-L101)?

---

**Screenshots and screen captures:**

<details>
<summary>API changelog update</summary>

![Screenshot from 2023-04-17 16-48-08](https://user-images.githubusercontent.com/63245456/232539073-ca9114d8-f9aa-471e-bd37-2e6ec6c4e8b7.png)
</details>

<details>
<summary>Send a message</summary>

[Current documentation](https://zulip.com/api/send-message)
<details>
<summary>Main endpoint description and usage examples</summary>

![Screenshot from 2023-04-17 16-50-21](https://user-images.githubusercontent.com/63245456/232538394-d2ed3339-3b7c-4f68-89dd-f5398c1f22d7.png)
![Screenshot from 2023-04-17 16-50-32](https://user-images.githubusercontent.com/63245456/232538398-492451a4-471a-4605-bc02-29875decb79a.png)
![Screenshot from 2023-04-17 16-50-40](https://user-images.githubusercontent.com/63245456/232538401-ba956f7a-9b0f-4f5c-b672-a306edb91c6b.png)
![Screenshot from 2023-04-17 16-50-47](https://user-images.githubusercontent.com/63245456/232538415-a9785037-5a4a-4249-8b1d-ef82f095fe85.png)
</details>
<details>
<summary>Parameters and response</summary>

![Screenshot from 2023-04-17 16-50-57](https://user-images.githubusercontent.com/63245456/232538419-25edea0c-d051-4b30-ac9e-ae255836535b.png)
![Screenshot from 2023-04-17 16-51-06](https://user-images.githubusercontent.com/63245456/232538421-708e15bf-1344-4966-a219-9adb0c162fc7.png)
</details>
</details>
<details>
<summary>Set typing status</summary>

[Current documentation](https://zulip.com/api/set-typing-status)
<details>
<summary>Main endpoint description</summary>

![Screenshot from 2023-04-17 16-49-12](https://user-images.githubusercontent.com/63245456/232540431-eb421fa7-c1de-4853-a0b8-f769eee22530.png)
</details>
<details>
<summary>Usage examples</summary>

![Screenshot from 2023-04-17 16-49-30](https://user-images.githubusercontent.com/63245456/232540157-7b0b9d8f-2e16-45dd-b0db-d07446fda0e7.png)
![Screenshot from 2023-04-17 16-49-36](https://user-images.githubusercontent.com/63245456/232540163-f4c557e4-1224-4d22-a2d8-71c265527a3c.png)
![Screenshot from 2023-04-17 16-49-44](https://user-images.githubusercontent.com/63245456/232540167-bd604c45-5ee6-491b-a909-12d114a8499a.png)
</details>
<details>
<summary>Parameters and response</summary>

![Screenshot from 2023-04-17 16-49-54](https://user-images.githubusercontent.com/63245456/232540171-d6a88467-b198-4309-932b-1bf93f9b60ed.png)
![Screenshot from 2023-04-17 16-50-02](https://user-images.githubusercontent.com/63245456/232540176-206d9437-f35a-4a75-b07a-7a9f67b2409d.png)
</details>
</details>
---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/lib/message.py:
```diff
@@ -173,6 +173,9 @@ class SendMessageRequest:
 
 
+# Used to check for valid string values for message `type` parameters.
+VALID_MESSAGE_TYPES = ["direct", "private", "stream"]
```
I wasn't sure if there was another place in the backend code where it would be better to define this constant so that it can be shared by both `zerver/views/message-send.py` and `zerver/views/typing.py`?

timabbott replied:
I suppose `Message.API_RECIPIENT_TYPES` is the other place I'd consider, but that might be confusing. I think we should call these "recipient types", not "message types", because we have considered making "message type" be a phase reserved for special types of messages, like polls and certain types of special Notification Bot messages.

laurynmm replied:
Okay. I moved this to `Message.API_RECIPIENT_TYPES` and added a comment. I used the comments for other `API_` constants in `models.py` as a jumping off point for that comment, but it may need some revision.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5502,25 +5502,33 @@ paths:
       tags: ["messages"]
       description: |
-        Send a stream or a private message.
+        Send a [stream](/help/streams-and-topics) or
+        a [direct](/help/direct-messages) message.
```
nit, and I see this is a pre-existing issue, but since we're looking at this sentence: this wording parses much more strongly for me as "send a stream, or send a direct message" than as "send a stream message, or send a direct message".

Probably better to just say "Send a [stream message](…) or a [direct message](…)".

timabbott replied:
Yeah that's a good improvement.

laurynmm replied:
Agreed. Updated in current version.

gnprice reviewed (commented):
Thanks! One drive-by nit.

timabbott commented:
> Did not update the python example for "direct". I wasn't sure if we wanted to update that in tandem with an update to zulip-python-api; see question below.

Yeah, we should do that after doing a `python-zulip-api` release to support "direct" and auto-translate it to "private" for older servers.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1154,7 +1154,7 @@ def check_send_private_message(
     sender: UserProfile,
     client: Client,
-    message_type_name: str,
+    message_type: str,
```
I think `recipient_type_name` would be a better variable name, since it's really a type of recipient, and `recipient_type` might refer to the 1/2/3 constants for `STREAM`, `HUDDLE`, etc.

laurynmm replied:
Updated the prep commit to use `recipient_type_name` and reviewed a `git-grep` of what's remaining in `/zerver` for `message_type`. They are all either part of the events change that we're not planning for this release or are unrelated to this part of the codebase.

```console
$ git grep -c message_type zerver/
zerver/actions/invites.py:1
zerver/actions/message_delete.py:5
zerver/actions/message_edit.py:1
zerver/actions/message_send.py:2
zerver/actions/typing.py:2
zerver/lib/event_schema.py:11
zerver/lib/events.py:1
zerver/migrations/0377_message_edit_history_format.py:3
zerver/openapi/zulip.yaml:14
zerver/tests/test_events.py:4
zerver/tests/test_realm_export.py:1
zerver/tests/test_retention.py:1
zerver/tests/test_service_bot_system.py:2
zerver/views/realm_export.py:1
zerver/webhooks/groove/fixtures/ticket_assigned__agent_and_group.json:1
zerver/webhooks/groove/fixtures/ticket_assigned__agent_only.json:1
zerver/webhooks/groove/fixtures/ticket_assigned__group_only.json:1
zerver/webhooks/groove/fixtures/ticket_assigned__no_one.json:1
zerver/webhooks/groove/fixtures/ticket_started.json:1
zerver/webhooks/groove/fixtures/ticket_state_changed.json:1
zerver/webhooks/pagerduty/view.py:10

```

timabbott commented on zerver/openapi/javascript_examples.js:
```diff
@@ -252,7 +252,7 @@ add_example("set_typing_status", "/typing:post", 200, async (client, console) =>
     };
 
-    // The user has started to type in the group PM with Iago and Polonius
+    // The user has started to type in the group DM with Iago and Polonius
```
This is a nit, but we've generally been trying to just spell these out, say "group direct message", in documentation, since we generally avoid acronyms as a project.

laurynmm replied:
Done in the updated version of these changes. Added a line break so that it renders clearly in the documentation.

timabbott commented on zerver/views/message_send.py:
```diff
@@ -211,6 +211,11 @@ def send_message_backend(
+    if message_type == "direct":
+        # For now, use "private" message type for direct messages.
+        # TODO: Use "direct" type here, as well as in events and
+        # message (created, schdeduled, drafts) objects/dicts.
```
A TODO here sounds reasonable. I think changing events is probably out of scope for this release; it's a more complex migration requiring a client capability.

laurynmm replied:
Yeah, that makes sense. Mostly trying to leave us in a good place with these changes for looking at that more complex change for after the release settles.

laurynmm commented:
**Updates**:
- Addressed review feedback above.
- Added one more prep commit for the refactor from using `message_type_name` or `message_type` to using `recipient_type_name` (where possible) for an instance where `recipient_type` was more appropriate.
- For the **Changes** note in `POST /typing`, I didn't include the second sentence about clients needing to use the modern convention because the default value changed from "private" to "direct", so I imagine most clients won't need to do anything and can keep using the default.

---

**Notes**:
- Follow-up when this is merged will be filing the `zulip-python-api` issue.

---

**Screenshots**:
Note that these are focused on areas that were changed due to review feedback.

<details>
<summary>API changelog feature level note</summary>

![Screenshot from 2023-04-18 19-31-15](https://user-images.githubusercontent.com/63245456/232859233-2b4b85c1-ec9b-4e8c-89cc-d572e669debe.png)

</details>
<details>
<summary>Send a message</summary>

![Screenshot from 2023-04-18 19-32-20](https://user-images.githubusercontent.com/63245456/232859074-a194999a-c8a3-46ec-88c5-9d950e4a1ed4.png)
![Screenshot from 2023-04-18 19-32-30](https://user-images.githubusercontent.com/63245456/232859088-48e240ee-c6e2-49bb-b27e-be336b217829.png)
</details>
<details>
<summary>Set typing status</summary>

![Screenshot from 2023-04-18 19-31-46](https://user-images.githubusercontent.com/63245456/232859358-72cea4d3-de4c-4bda-bd21-33e9a266c61f.png)
![Screenshot from 2023-04-18 19-32-05](https://user-images.githubusercontent.com/63245456/232859364-dce6556d-6ac3-47da-bb97-f22c16b3a85c.png)
</details>

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -5502,25 +5502,33 @@ paths:
+            **Changes**: In Zulip 7.0 (feature level 174), added `"direct"` to
+            indicate that the message to be sent is a direct message to specific
+            users and not a stream message. `"private"` is now a deprecated alias
+            for the `"direct"` option to be removed in a future release.
```
Thanks for the revision here. I modified it slightly for the main API changelog note and the note in the `/typing` endpoint.

laurynmm commented on zerver/views/message_send.py:
```diff
@@ -193,7 +193,7 @@ def handle_deferred_message(
     request: HttpRequest,
     user_profile: UserProfile,
-    message_type_name: str = REQ("type"),
+    req_type: str = REQ("type", str_validator=check_string_in(Message.API_RECIPIENT_TYPES)),
```
I went with the `req_` format we use in a few places here to clarify this is a parameter from the request.

timabbott replied:
That works for me!

timabbott commented:
> I didn't include the second sentence about clients needing to use the modern convention because the default value changed from "private" to "direct", so I imagine most clients won't need to do anything and can keep using the default.

Sounds good -- we actually haven't finished implementing typing notifications for stream messages in the web frontend yet.

timabbott commented on zerver/views/message_send.py:
```diff
@@ -210,12 +210,14 @@ def send_message_backend(
     tz_guess: Optional[str] = REQ("tz_guess", default=None, documentation_pending=True),
     time: Optional[float] = REQ(default=None, converter=to_float, documentation_pending=True),
 ) -> HttpResponse:
+    recipient_type_name = message_type_name
```
This can be done a bit more cleanly be using the first positional argument to `REQ`, to not have set `message_type_name` at all.

timabbott replied:
Ahh, but it gets replaced in the next commit anyway, yeah this is pretty reasonable as an intermediate state before we redo the internals plumbing to support `"direct"` rather than `"private"`.

timabbott commented:
Merged, huge thanks for doing this migration @laurynmm!
